### PR TITLE
Add linting for mismatching build configurations in a workspace

### PR DESCRIPTION
**Pull request by kwridan**
Resolves https://github.com/tuist/tuist/issues/467

### Short description 📝

It can be easy to miss specifying a project build configuration in large workspaces. This can lead to build errors or worse an incorrectly configured built project!

### Solution 📦

Add a linter to flag any missing or mismatching build configurations.

### Implementation 👩‍💻👨‍💻

- All projects' build configurations are now validated against the top level projects
- In the event any are missing or are mismatching (e.g. have a mismatching variant) the linter will now flag them
- Dependencies with more configurations defined won't get flagged (as those lower level dependencies can be shared in other workspaces which could have different configurations)

Example:

```
$ tuist generate
The following issues have been found:
  - The project 'Framework2' has missing or mismatching configurations. It has [Debug (debug), Release (release)], other projects have  [Beta (release), Debug (debug), Release (release)].
```

- [x] Add lint step
- [x] Update change log

### Test Plan 🛠

- Run `tuist generate` within `fixtures/ios_app_with_multi_configs`
- Comment out any of the build configurations defined in `App`, `Framework1` or `Framework2`
- Re-run `tuist generate`
- Verify a lint warning is displayed

### CODE REVIEW

File: Sources/TuistGenerator/Linter/GraphLinter.swift [hunk 1]
```diff
@@ -161,11 +186,11 @@ class GraphLinter: GraphLinting {
...
-//            LintableTarget(platform: .iOS, product: .appExtension),
-//            LintableTarget(platform: .iOS, product: .messagesExtension),
-//            LintableTarget(platform: .iOS, product: .stickerPack),
-//            LintableTarget(platform: .watchOS, product: .watch2App),
-//            LintableTarget(platform: .watchOS, product: .watchApp),
+            //            LintableTarget(platform: .iOS, product: .appExtension),
+            //            LintableTarget(platform: .iOS, product: .messagesExtension),
+            //            LintableTarget(platform: .iOS, product: .stickerPack),
+            //            LintableTarget(platform: .watchOS, product: .watch2App),
+            //            LintableTarget(platform: .watchOS, product: .watchApp),
```
Comment (ollieatkinson) [inline Sources/TuistGenerator/Linter/GraphLinter.swift]: Just double check you're happy with this new alignment, I think by formatting the file it's indented them by accident.

File: Tests/TuistGeneratorTests/Linter/GraphLinterTests.swift [hunk 2]
```diff
@@ -247,4 +247,119 @@ final class GraphLinterTests: XCTestCase {
...
+        // Then
+        XCTAssertEqual(result, [
+            LintingIssue(reason: "The project 'ProjectC' has missing or mismatching configurations. It has [Beta (release), Debug (release), Release (release), Testing (release)], other projects have [Beta (release), Debug (debug), Release (release), Testing (debug)]",
+                         severity: .warning),
+        ])
+    }
+
+    func test_lint_doesNotFlagDependenciesWithExtraConfigurations() throws {
+        // Lower level dependencies could be shared by projects in different workspaces as such
+        // it is ok for them to contain more configurations than the entry node projects
```
Comment (ollieatkinson) [inline Tests/TuistGeneratorTests/Linter/GraphLinterTests.swift]: :+1:

File: Sources/TuistGenerator/Linter/GraphLinter.swift (hunk 1 above)
Comment (kwridan) [reply Sources/TuistGenerator/Linter/GraphLinter.swift]: it was swiftformat - will push a revert, hopefully won't get flagged as an unformatted PR 🤐

Comment (kwridan) [reply Sources/TuistGenerator/Linter/GraphLinter.swift]: ouch, will have to re-apply those change

Screen Shot 2019-08-03 at 11 55 38 AM

File: Sources/TuistGenerator/Linter/GraphLinter.swift [hunk 3]
```diff
@@ -148,6 +149,30 @@ class GraphLinter: GraphLinting {
         return [issue]
     }
 
+    private func lintMismatchingConfigurations(graph: Graphing) -> [LintingIssue] {
+        let entryNodeProjects = graph.entryNodes.compactMap { $0 as? TargetNode }.map { $0.project }
+
+        let knownConfgiruations = entryNodeProjects.reduce(into: Set()) {
```
Comment (pepibumur) [inline Sources/TuistGenerator/Linter/GraphLinter.swift]: Typo: knownConfgiruations

**Pull request merged**